### DISCUSSION TITLE
[FEATURE REQUEST] improvement of windows kb uninstallation

### DISCUSSION THREAD

**Discussion by sfonteneau** [saltstack/salt#67460] [Feature Requests]
I am working on a solution that matches yours ([wapt.fr](https://www.wapt.fr/en/doc/)). 
But nothing stops us from joining forces!

For : 
https://github.com/saltstack/salt/blob/daef7613bc359822a4f04ffdd7a91a9edcac7c80/salt/utils/win_update.py#L986

I'm sharing with you an improvement that I found to uninstall Windows KB:

```
import xml.etree.ElementTree as ET
import glob
import run
import disable_file_system_redirection

def get_kb_dism_name():

    dict_dism_kb={}

    with disable_file_system_redirection():
        for f in glob.glob(r'C:\Windows\servicing\Packages\*.mum'):
            identifier = None
            dism_name = None
            tree =  ET.parse(f)
            root =  tree.getroot()
            for child in root:
                if child.tag == '{urn:schemas-microsoft-com:asm.v3}assemblyIdentity':
                    dism_name = child.attrib['name'] + '~' + child.attrib['publicKeyToken'] + "~" + child.attrib['processorArchitecture'] + '~~' + child.attrib['version']
                if child.tag == '{urn:schemas-microsoft-com:asm.v3}package':
                    identifier = child.attrib['identifier']

            if dism_name and identifier :
                dict_dism_kb[dism_name] = identifier

        dict_final_dism_kb = {}
        data = run('dism /Online /Get-Packages')
        for entry in data.splitlines():
            if entry.split(' : ',1)[-1] in dict_dism_kb:
                name_entry = entry.split(' : ',1)[-1]
                dict_final_dism_kb[dict_dism_kb[name_entry].lower()] = name_entry

    return dict_final_dism_kb

def test():

    data_dism=get_kb_dism_name()

    dism_name = data_dism['kb5005699']

    print('dism /Online /Remove-Package /PackageName:%s /Quiet /NoRestart' % dism_name)
    
    
test()

```

With this we can find the dism name of a kb even if the kb name is not in dism name.

See you soon !

**Comment by welcome** [comment]
Hi there! Welcome to the Salt Community! Thank you for making your first contribution. We have a lengthy process for issues and PRs. Someone from the Core Team will follow up as soon as possible. In the meantime, here’s some information that may help as you continue your Salt journey.
Please be sure to review our [Code of Conduct](https://github.com/saltstack/salt/blob/master/CODE_OF_CONDUCT.md). Also, check out some of our community resources including:

  - [Community Wiki](https://github.com/saltstack/community/wiki)
  - [Salt’s Contributor Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html)
  - [Join our Community Slack](https://via.vmw.com/salt-slack)
  - [IRC on LiberaChat](https://web.libera.chat/#salt)
  - [Salt Project YouTube channel](https://www.youtube.com/channel/UCpveTIucFx9ljGelW63-BWg)
  - [Salt Project Twitch channel](https://www.twitch.tv/saltprojectoss)

There are lots of ways to get involved in our community. Every month, there are around a dozen opportunities to meet with other contributors and the Salt Core team and collaborate in real time. The best way to keep track is by subscribing to the Salt Community Events Calendar.
If you have additional questions, email us at saltproject@vmware.com. We’re glad you’ve joined our community and look forward to doing awesome things with you!
